index.js

```diff
@@ -1,189 +1,317 @@
+ * @name hasProperty
  * @api public
  */
 
-exports.set = function(obj, path, val) {
-  var parsed = exports.parse(path);
-  setPathValue(parsed, val, obj);
+var literals = {
+  'number': Number,
+  'string': String,
 };
+function hasProperty(name, obj) {
+  var objType = type(obj);
+  // Bad Object, obviously no props at all
+  if (objType === 'null' || objType === 'undefined') {
+    return false;
+  }
 
-/*!
+  // The `in` operator does not work with certain literals
+  // box these before the check
+  if (literals[objType] && typeof obj !== 'object') {
+    obj = new literals[objType](obj);
+  }
+
+  return name in obj;
+}
+
+/* !
+ * ## parsePath(path)
+ *
  * Helper function used to parse string object
- * paths. Use in conjunction with `getPathValue`.
+ * paths. Use in conjunction with `internalGetPathValue`.
  *
- *  var parsed = parsePath('myobject.property.subprop');
+ *      var parsed = parsePath('myobject.property.subprop');
  *
  * ### Paths:
  *
- * * Can be as near infinitely deep and nested
+ * * Can be infinitely deep and nested.
  * * Arrays are also valid using the formal `myobject.document[3].property`.
+ * * Literal dots and brackets (not delimiter) must be backslash-escaped.
  *
  * @param {String} path
- * @returns {Array} parsed
+ * @returns {Object} parsed
+ * @api private
  */
 
-exports.parse = function(path) {
-  var str = (path || '').replace(/\[/g, '.[');
+function parsePath(path) {
+  var str = path.replace(/([^\\])\[/g, '$1.[');
   var parts = str.match(/(\\\.|[^.]+?)+/g);
-  var re = /\[(\d+)\]$/;
-  var ret = [];
-  var mArr = null;
-
-  for (var i = 0, len = parts.length; i < len; i++) {
-    mArr = re.exec(parts[i]);
-    ret.push(mArr ? { i: parseFloat(mArr[1]) } : { p: parts[i] });
-  }
+  return parts.map(function mapMatches(value) {
+    var regexp = /^\[(\d+)\]$/;
+    var mArr = regexp.exec(value);
+    var parsed = null;
+    if (mArr) {
+      parsed = { i: parseFloat(mArr[1]) };
+    } else {
+      parsed = { p: value.replace(/\\([.\[\]])/g, '$1') };
+    }
 
-  return ret;
-};
+    return parsed;
+  });
+}
 
-/*!
- * Companion function for `parsePath` that returns
+/* !
+ * ## internalGetPathValue(parsed, obj)
+ *
+ * Helper companion function for `.parsePath` that returns
  * the value located at the parsed address.
  *
- *  var value = getPathValue(parsed, obj);
+ *      var value = getPathValue(parsed, obj);
  *
  * @param {Object} parsed definition from `parsePath`.
  * @param {Object} object to search against
+ * @param {Number} object to search against
  * @returns {Object|Undefined} value
+ * @api private
  */
 
-function getPathValue(parsed, obj) {
-  var tmp = obj;
-  var res;
+function internalGetPathValue(parsed, obj, pathDepth) {
+  var temporaryValue = obj;
+  var res = null;
+  pathDepth = (typeof pathDepth === 'undefined' ? parsed.length : pathDepth);
 
-  for (var i = 0, l = parsed.length; i < l; i++) {
+  for (var i = 0, pathPart = pathDepth; i < pathPart; i++) {
```

Comment:
What's the purpose of `pathPart`? Why can't `pathDepth` be used directly instead?